nursery_store.go

```diff
@@ -0,0 +1,1377 @@
+			chanBucket := chanIndex.Bucket(chanBytes)
+			if chanBucket == nil {
+				continue
+			}
+
+			// All of the outputs of interest will start with the
+			// "pscl" prefix. So, we will perform a prefix scan of
+			// the channel bucket to efficiently enumerate all the
+			// desired outputs.
+			c := chanBucket.Cursor()
+
+			// Seek and iterate over all outputs starting with the
+			// prefix "pscl".
+			pfxOutputKey, kidBytes := c.Seek(psclPrefix)
+			for bytes.HasPrefix(pfxOutputKey, psclPrefix) {
+
+				// Deserialize each output as a kidOutput, since
+				// this should have been the type that was
+				// serialized when it was written to disk.
+				var psclOutput kidOutput
+				psclReader := bytes.NewReader(kidBytes)
+				err := psclOutput.Decode(psclReader)
+				if err != nil {
+					return err
+				}
+
+				// Add the deserialized output to our list of
+				// preschool outputs.
+				kids = append(kids, psclOutput)
+
+				// Advance to the subsequent key-value pair of
+				// the prefix scan.
+				pfxOutputKey, kidBytes = c.Next()
+			}
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return kids, nil
+}
+
+// ForChanOutputs iterates over all outputs being incubated for a particular
+// channel point. This method accepts a callback that allows the caller to
+// process each key-value pair. The key will be a prefixed outpoint, and the
+// value will be the serialized bytes for an output, whose type should be
+// inferred from the key's prefix.
+// NOTE: The callback should be not modify the provided byte slices and is
+// preferably non-blocking.
+func (ns *nurseryStore) ForChanOutputs(chanPoint *wire.OutPoint,
+	callback func([]byte, []byte) error) error {
+
+	return ns.db.View(func(tx *bolt.Tx) error {
+		return ns.forChanOutputs(tx, chanPoint, callback)
+	})
+}
+
+// IsMatureChannel determines the whether or not all of the outputs in a
+// particular channel bucket have been marked as graduated.
+func (ns *nurseryStore) IsMatureChannel(chanPoint *wire.OutPoint) (bool, error) {
+	var isMature bool
+	if err := ns.db.View(func(tx *bolt.Tx) error {
+		// Iterate over the contents of the channel bucket, computing
+		// both total number of outputs, and those that have the grad
+		// prefix.
+		var nOutputs, nGrads int
+		if err := ns.forChanOutputs(tx, chanPoint, func(pfxKey, _ []byte) error {
+			if string(pfxKey[:4]) == string(gradPrefix) {
+				nGrads++
+			}
+			nOutputs++
+
+			return nil
+
+		}); err != nil {
+			return err
+		}
+
+		// Channel is mature if all outputs are graduated.
+		if nGrads == nOutputs {
+			isMature = true
+		}
+
+		utxnLog.Infof("ChannelID(%v) found %d graduated outputs out of %d",
+			chanPoint, nGrads, nOutputs)
+
+		return nil
+	}); err != nil {
+		return false, err
+	}
+
+	return isMature, nil
+}
+
+// RemoveChannel channel erases all entries from the channel bucket for the
+// provided channel point.
+func (ns *nurseryStore) RemoveChannel(chanPoint *wire.OutPoint) error {
```

Comment:
Seems a little odd that this doesn't also clean up the height index. If that is desired, it should probably be specifically be called out in the method doc.

Reply:
Still fleshing out the exact pruning strategy for the height index, but you bring up a good point in that we should definitely attempt to remove all outputs in the channel bucket from the height index before deleting the channel bucket itself. In its current form, this may only be possible for kindergarten outputs, since we drop some information from the baby txns when they move to kindergarten. However, height buckets are opportunistically cleaned up after the confirmation of a htlc timeout or kindergarten sweep txn, and after some pruning depth has been reached. +1 for adding this as an additional safety precaution